---
title: "CCBR-1066 Primary Analysis"
output: html_document
editor_options: 
  chunk_output_type: console
---
# Code info
Runs GSEA and ORA with Pipeliner Output for Human data
Developed from code by Vishal Koparde

#####################################
# Setup
#####################################
Load libraries
```{r, include=FALSE}
library(VennDiagram)
library(scales)
library(dplyr)
library("tidyverse")
library("stats")
library("affy")
library("DESeq2")
library("edgeR")
library("DT")
library("ggplot2")
library("EnhancedVolcano")
library("msigdbr")
library("enrichplot")
library("clusterProfiler")
library("ggpubr")
library("pheatmap")
library(RColorBrewer)
library(plotly)
```

Set directories
```{r setup, include=FALSE}
analysis_dir ="~/../../Volumes/ccbr1109/analysis/"
deg_dir="~/../../Volumes/ccbr1109/rnaseq/"

deg_merged_input=paste0(analysis_dir,"deg_data/merged_deg_fc_DESeq2_20210811.txt")
kmeans_input = paste0(analysis_dir,"kmeans/Kmeans.csv")
```

Set global variables
```{r}
p_val = 0.05
species = "mouse" #Homo sapiens
fc = 1.5

#set deg list
deg_list = c(paste0(deg_dir, "DEG_GCN1_SH2-SCR_1_1/DESeq2_DEG_GCN1_SH2-SCR_all_genes.txt"),
             paste0(deg_dir, "DEG_HNPK_SH1-SCR_1_1/DESeq2_DEG_HNPK_SH1-SCR_all_genes.txt"),
             paste0(deg_dir, "DEG_HNPK_SH2-SCR_1_1/DESeq2_DEG_HNPK_SH2-SCR_all_genes.txt"))

compare_list = c("GCN1_SH2.SCR","HNPK_SH1.SCR","HNPK_SH2.SCR")
cat_list = c("GCN1_SH2","HNPK_SH1","HNPK_SH2")

#kmeans clusters
cluster_list=c("B","C")

version_id="v93"
```

#####################################
# pCOA plots
#####################################
```{r, echo=FALSE,webgl=TRUE,message=FALSE, warning=FALSE}
generate_pca_plots<-function(file.in){
  #read sample info
  sampledf=read.delim(paste0(deg_dir,"STAR_files/","sampletable.txt"))
  
  #read DEG info
  degdf = read.delim(paste0(deg_dir,"DEG_ALL/","RawCountFile_RSEM_genes_filtered.txt"),row.names=1)
  colnames(degdf)=as.character(sampledf[,4])
  sampleFiles=as.character(sampledf[,2])
  
  ## read annotation file
  condition = as.factor(sampledf$condition)
  y = DGEList(counts=degdf,group=condition)
  y <- calcNormFactors(y,method="TMM")
  
  #add group
  Group <- factor(sampledf$condition)
  if ( length(levels(Group)) == 1 ) {
    design=~1
    v1 <- voom(as.matrix(degdf),design=NULL,plot=FALSE,normalize="quantile")
  } else {
    design=model.matrix(~0+Group)
    v1 <- voom(as.matrix(degdf),design,plot=FALSE,normalize="quantile")
  }
  ddsHTSeq<-DESeqDataSetFromMatrix(countData=degdf,colData=sampledf, design=design)
  dds<-DESeq(ddsHTSeq)
  dds.ndata=as.data.frame(counts(dds,normalized=TRUE))
  
  ## estimating common and tagwise dispersions
  y <- estimateCommonDisp(y)
  y <- estimateTagwiseDisp(y)
  
  ylog2=cpm(y,log=TRUE,normalized.lib.sizes=TRUE,prior.count=0.5)
  rawlog2= cpm(y,log=TRUE,normalized.lib.sizes=FALSE,prior.count=0.5)
  rld <- rlogTransformation(dds, blind=TRUE)
  rldm=assay(rld)
  colnames(rldm)=colnames(degdf)
  
  #deseq2
  deseq2.edf=as.matrix(rldm)
  deseq2.tedf= t(deseq2.edf)
  deseq2.tedf=deseq2.tedf[,apply(deseq2.tedf,2,var)!= 0]
  deseq2.pca=prcomp(deseq2.tedf,scale.=T)
  deseq2.tedf1 = data.frame(deseq2.tedf)
  Phenotype=sampledf$condition
  cell_rep=sampledf$label
  deseq2.tedf1$group = as.factor(Phenotype)
  deseq2.pc1 = round(deseq2.pca$sdev[1]^2/sum(deseq2.pca$sdev^2)*100,2)
  deseq2.pc2 = round(deseq2.pca$sdev[2]^2/sum(deseq2.pca$sdev^2)*100,2)
  deseq2.pc3 = round(deseq2.pca$sdev[3]^2/sum(deseq2.pca$sdev^2)*100,2)
  
  #pca plotting
  pcafactor = as.factor(sampledf$condition)
  col <- brewer.pal(nlevels(pcafactor), "Paired")
  
  deseq2df = as.data.frame(deseq2.pca$x[,1:3])
  deseq2df$Groups=sapply(strsplit(rownames(deseq2df),".", fixed=TRUE), `[`, 1)
  
  p1 = ggplot(deseq2df, aes(x=PC1, y=PC2, color=Groups)) + 
      geom_point() + theme(legend.position="none")
  p2 = ggplot(deseq2df, aes(x=PC1, y=PC3, color=Groups)) + 
      geom_point() + theme(legend.position="none")
  p3 = ggplot(deseq2df, aes(x=PC2, y=PC3, color=Groups)) + 
      geom_point()
  pf = ggarrange(p1, p2, p3,
                 labels = c("A", "B", "C"),
                 ncol = 2, nrow = 2, common.legend = TRUE)
  print(pf)
  ggsave(filename = file.in, 
           height = 8.90, width = 12.80, device = "png", plot = pf)
}

generate_pca_plots(paste0(analysis_dir,"pca/","pca_plot.png"))

#render 3D plot
# p <- plot_ly(as.data.frame(deseq2.pca$x[,1:3]),
#              x = ~PC1, y = ~PC2, z = ~PC3, color = pcafactor,
#              colors = col, hoverinfo="text", hovertext = ~sampledf$label) %>%
#   add_markers() %>%
#   layout(title = "DESeq2 PCA plot",
#          scene = list(xaxis = list(title = paste0("PC1 (",deseq2.pc1,"%)")),
#                       yaxis = list(title = paste0("PC2 (",deseq2.pc2,"%)")),
#                       zaxis = list(title = paste0("PC3 (",deseq2.pc3,"%)"))))
#p
```

#####################################
# ORA / GSA on DESEQ2 DEG
#####################################
Create gene venn diagrams
```{r}
create_venn_diagrams <-function(cat.in,out_dir){
  #create output dir if needed
  dir.create(file.path(out_dir), showWarnings = FALSE)
  
  #read in DEG cvs file
  deg_df = read.csv(deg_merged_input,sep="\t")

  #create gene lists
  lev = 1
  for (c_variable in compare_list){
    fdr_variable = paste(c_variable,".fdr",sep="")
    fc_variable = paste(c_variable,".fc",sep="")
  
    name_up = paste0("gene_up_",lev)
    name_down = paste0("gene_down_",lev)
    
    genes_up = lapply(deg_df[deg_df[[fdr_variable]]<=p_val & deg_df[[fc_variable]]<=-fc,]$gene, function(x) x[!is.na(x)])
    genes_down = lapply(deg_df[deg_df[[fdr_variable]]<=p_val & deg_df[[fc_variable]]>=fc,]$gene, function(x) x[!is.na(x)])
      
    assign(name_up,genes_up)#positive = KO upregulated
    assign(name_down,genes_down)#negative = upregulated over KO
    
    #save gene dfs
    write.csv(deg_df[deg_df[[fdr_variable]]<=p_val & deg_df[[fc_variable]]<=-fc,]%>% 
                 filter(if_all(everything(), ~ !is.na(.x))),
              paste0(out_dir,"genes_up_",c_variable,".csv"))
    write.csv(deg_df[deg_df[[fdr_variable]]<=p_val & deg_df[[fc_variable]]>=fc,]%>% 
                 filter(if_all(everything(), ~ !is.na(.x))),
              paste0(out_dir,"genes_down_",c_variable,".csv"))
    write.csv(deg_df[deg_df[[fdr_variable]]>=p_val | deg_df[[fc_variable]]<=fc | deg_df[[fc_variable]]>=-fc,]%>% 
                 filter(if_all(everything(), ~ !is.na(.x))),
              paste0(out_dir,"genes_nonde_",c_variable,".csv"))
    lev=lev+1
  }

  for (num in 1:length(compare_list)){
    print(compare_list[num])
    print(paste0("--UP: ",length(get(paste0("gene_up_",num)))))
    print(paste0("--DOWN: ",length(get(paste0("gene_down_",num)))))
  }
  
  # ven for upreglated
  venn.diagram(
    x = list(gene_up_1, gene_up_2, gene_up_3),
    category.names = cat.in,
    filename = paste0(out_dir,"venn_up.png"),
    output = TRUE , imagetype="png" ,
    height = 480 , width = 480 , resolution = 300,
    compression = "lzw",lwd = 1,
    col=c("#440154ff", '#21908dff', '#fde725ff'),
    fill = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3), alpha('#fde725ff',0.3)),
    cex = 0.5, fontfamily = "sans",
    cat.cex = 0.3, cat.default.pos = "outer",
    cat.pos = c(-27, 27, 135), cat.dist = c(0.055, 0.055, 0.085),
    cat.fontfamily = "sans", cat.col = c("#440154ff", '#21908dff', '#fde725ff'),
    rotation = 1
  )
  # ven for downreglated
  venn.diagram(
    x = list(gene_down_1, gene_down_2, gene_down_3),
    category.names = cat.in,
    filename = paste0(out_dir,"venn_down.png"),
    output = TRUE , imagetype="png" ,
    height = 480 , width = 480 , resolution = 300,
    compression = "lzw",lwd = 1,
    col=c("#440154ff", '#21908dff', '#fde725ff'),
    fill = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3), alpha('#fde725ff',0.3)),
    cex = 0.5, fontfamily = "sans",
    cat.cex = 0.3, cat.default.pos = "outer",
    cat.pos = c(-27, 27, 135), cat.dist = c(0.055, 0.055, 0.085),
    cat.fontfamily = "sans", cat.col = c("#440154ff", '#21908dff', '#fde725ff'),
    rotation = 1
  )
}

create_venn_diagrams(cat_list,
                     paste0(analysis_dir,"venn/"))
```

ORA/GSEA Functions
```{r functions, include=FALSE}
#read in df
read_deg_file<-function(fn){
  deg=as.data.frame(read.csv(fn,header=TRUE,sep="\t"))
# fix infinity issues
  if (min(deg$gsea_ranking_score)==-Inf){
    newmin=min(deg[!(deg$gsea_ranking_score==-Inf),]$gsea_ranking_score)
    deg[deg$gsea_ranking_score==-Inf,]$gsea_ranking_score=2*newmin
  }
  if (max(deg$gsea_ranking_score)==Inf){
    newmax=max(deg[!(deg$gsea_ranking_score==Inf),]$gsea_ranking_score)
    deg[deg$gsea_ranking_score==Inf,]$gsea_ranking_score=2*newmax
  }
  return(deg)
}

#format DEG gene list
deg_gene_list<-function(deg){
  gl=as.data.frame(deg$gsea_ranking_score)
  gl$GN=deg$gene
  colnames(gl)=c("Rank","GeneName")
  gl$absRank=abs(gl$Rank)
  gl=gl[order(gl$absRank,decreasing = TRUE),]
  gl=gl[match(unique(gl$GeneName),gl$GeneName),]
  geneList=gl$Rank
  names(geneList)=as.character(gl$GeneName)
  geneList <- sort(geneList, decreasing = TRUE)
  return(geneList)
}

#annotated gene info
annotate_genes <-function(type.in,sub.type){
  
  if(type.in=="c2"){
    c=msigdbr(species = species, 
              category = "C2", 
              subcategory = sub.type) %>% 
      dplyr::select(gs_name,gene_symbol)
  } else{
    c=msigdbr(species = species,  
              category = "C5", 
              subcategory = sub.type) %>% dplyr::select(gs_name,gene_symbol)
  }
  return(c)
}

#plot gsea/ora 
gsea_ora_plot <- function(gl,t2g,ttl,fn,type.in){
  
  #run enricher or gsea
  if(type.in == "gsea"){
    print("Running gsea")
    result=GSEA(geneList = gl,TERM2GENE = t2g, eps = 0, pvalueCutoff = 1)
  } else{
    print("Running ora")
    gl_clean =vapply(strsplit(gl,";"), `[`, 1, FUN.VALUE=character(1)) #kmeans list can have concatonated
    #genes separated by ; - only want the first gene listed
    result=enricher(gene=gl_clean, TERM2GENE=t2g, pvalueCutoff = 1)
  }
  
  #convert results to df and save
  resultdf=as.data.frame(result)
  write.table(resultdf,file=fn,quote=FALSE,sep="\t",row.names = FALSE,col.names = TRUE)
  resultdf$p.adjust=as.factor(round(resultdf$p.adjust,5))
  
  #if the n rows are 0, no sig results
  if(nrow(resultdf)==0){
    print(paste0("----No sig results for ",type.in))
    p1 = ggparagraph(
      paste0("\n\nNo Sig Results for ", type.in),
      color = NULL,
      size = 20,
      face = "bold",
      family = NULL,
      lineheight = NULL
    )
  #plot top 5 results  
  } else if(type.in=="ora"){
    # p1 = ggplot(resultdf)+
    #       geom_point(mapping = aes(x=GeneRatio, y=reorder(Description, -p.adjust), size=Count, color=p.adjust),
    #                    data = head(resultdf[order(resultdf$p.adjust),], n = 5)) +
    #       labs(y ="", x = "GeneRatio", color = "p.adjust") +
    #       theme(axis.text=element_text(size=8)) +
    #       ggtitle(ttl) +
    #       guides(size=FALSE) + 
    #       scale_fill_brewer(palette="RdBu") + 
    #       theme(panel.background = element_rect(fill = 'white', colour = 'black'))
    p1 = dotplot(result,title=ttl,font.size = 8, showCategory=10)
  } else{
    #if gsea then facet by the sign
    p1 = dotplot(result,title=ttl,font.size = 8, showCategory=5, split=".sign")+facet_grid(.~.sign)
  }
  return(p1)
}

#save all plots
save_plots<-function(p1,p2,file.in,p3="",p4=""){
  addSmallLegend <- function(myPlot, pointSize = 1, textSize = 6, spaceLegend = 0.1) {
    myPlot +
        guides(shape = guide_legend(override.aes = list(size = pointSize)),
               color = guide_legend(override.aes = list(size = pointSize))) +
        theme(legend.title = element_text(size = textSize), 
              legend.text  = element_text(size = textSize),
              legend.key.size = unit(spaceLegend, "lines"))
  }
  
  if(p3==""){
    pf = ggarrange(addSmallLegend(p1),
                   addSmallLegend(p2),
                 labels = c("A", "B"),
                 ncol = 1, nrow = 2)
    print(pf)
    ggsave(filename = file.in, 
           height = 8.90, width = 12.80, device = "png", plot = pf)
  } else if (p4==""){
    pf = ggarrange(addSmallLegend(p1),
                   addSmallLegend(p2),
                   addSmallLegend(p3),
                 labels = c("A", "B", "C"),
                 ncol = 1, nrow = 3)
    print(pf)
    ggsave(filename = file.in, 
           height = 8.90, width = 12.80, device = "png", plot = pf)
  } else{
    pf = ggarrange(addSmallLegend(p1),
                   addSmallLegend(p2),
                   addSmallLegend(p3),
                   addSmallLegend(p4),
                 labels = c("A", "B","C","D"),
                 ncol = 1, nrow = 4)
    print(pf)
    ggsave(filename = file.in, 
           height = 8.90, width = 12.80, device = "png", plot = pf)
  }
}

#processing function
main_function<-function(deg_file,out_dir,contrast_id){
  
  # read deg
  deg=read_deg_file(deg_file)
  
  #create output dir if needed
  dir.create(file.path(out_dir), showWarnings = FALSE)

  #subset significant genes for ORA
  siggenes=deg[deg$fdr <= p_val & (deg$fc < (-1*fc) | deg$fc > fc),]
  sigGeneList=siggenes$gene
  
  #set genelist for GSEA
  geneList=deg_gene_list(deg)
  
  #Run C2 plots
  c_list = c("BIOCARTA","KEGG","REACTOME","WIKIPATHWAYS")
  for (cid in c_list){
    print(paste0("--Results for: ",cid))
    #p1=""
    p1 = gsea_ora_plot(gl=geneList,
                      t2g=annotate_genes("c2",cid),
                      ttl=paste0("GSEA:C2:",cid," - ",contrast_id),
                       fn=paste0(out_dir,contrast_id,".c2.",cid,".gsea.results.txt",sep=""),
                      "gsea")
    p2 = gsea_ora_plot(gl=sigGeneList,
                       t2g=annotate_genes("c2",cid),
                       ttl=paste0("ORA:C2:",cid," - ",contrast_id),
                        fn=paste0(out_dir,contrast_id,".c2.",cid,".ora.results.txt",sep=""),
                       "ora")
    save_plots(p2,p1,paste0(out_dir, contrast_id, "/", contrast_id, ".c2.",cid,".dotplot.png"))
  }
  
  #Run C5 plots
  c_list = c("GO:MF","GO:BP","GO:CC")
  for (cid in c_list){
    print(paste0("--Results for: ",cid))
    p1 = gsea_ora_plot(gl=geneList,
                      t2g=annotate_genes("c5",cid),
                     ttl=paste0("GSEA:C5:",cid," - ",contrast_id),
                      fn=paste0(out_dir,contrast_id,".c5",cid,".gsea.results.txt",sep=""),
                      "gsea")
    p2 = gsea_ora_plot(gl=sigGeneList,
                       t2g=annotate_genes("c5",cid),
                       ttl=paste0("ORA:c5:",cid," - ",contrast_id),
                        fn=paste0(out_dir,contrast_id,".c5.",cid,".ora.results.txt",sep=""),
                       "ora")
    save_plots(p2,p1,paste0(out_dir, contrast_id, "/", contrast_id, ".c5.",cid,".dotplot.png"))
  }
}
```

ORA/GSEA Run
```{r setup, include=FALSE}
#paths
out_dir = paste0(analysis_dir,"ora_vs_gsea/")

#GCN1_SH2 vs SCR
print("**GCN1_SH2 vs SCR")
main_function(deg_file = deg_list[1], 
     out_dir = paste0(out_dir,compare_list[1],"/"),
     contrast_id = compare_list[1])

#HNPK_SH1 vs SCR
print("**HNPK_SH1 vs SCR")
main_function(deg_file = deg_list[2], 
     out_dir = paste0(out_dir,compare_list[2],"/"),
     contrast_id = compare_list[2])

#HNPK_SH2 vs SCR
print("**HNPK_SH2 vs SCR")
main_function(deg_file = deg_list[3], 
     out_dir = paste0(out_dir,compare_list[3],"/"),
     contrast_id = compare_list[3])
```

GSEA Heatmap Run
```{r}
#heatplot function
gsea_plus_heatplot<-function(deg_list, out_dir, contrast_list, annotation_df){
  #for each annotation in df, create heatmap for each contrast input
  for(rownum in rownames(annotation_df)){
    #start deg counter at 1
    counter_deg=1
    
    #create db
    anno_db = msigdbr(species = species,category = annotation_df[rownum,"ID"], 
                             subcategory = annotation_df[rownum,"SUB"]) %>% dplyr::select(gs_name,gene_symbol)
    
    #compare each contrast at this level
    print(paste0("**Heatplots for ", annotation_df[rownum,"file_name"]))
    
    for(contrast_id in contrast_list){
      print(paste0("--",contrast_id))
      
      # read deg
      deg=read_deg_file(deg_list[counter_deg])
      
      #create output dir if needed for pval and NES output
      dir.create(file.path(out_dir), showWarnings = FALSE)
      dir.create(file.path(paste0(out_dir,"pvalue")), showWarnings = FALSE)
      dir.create(file.path(paste0(out_dir,"NES")), showWarnings = FALSE)
    
      #set genelist for ORA
      geneList=deg_gene_list(deg)
      
      ## pvalue
      #pull top paths
      tmp = GSEA(geneList = geneList,TERM2GENE = anno_db,eps = 0, pvalueCutoff = 1) #run GSEA
      
      if(nrow(tmp)>0){
        tmp_df = as.data.frame(tmp@result) #create df from results
        tmp_df = tmp_df[order(tmp_df$pvalue),] #order by pvalue
        tmp_df = tmp_df[1:5,c("ID","pvalue")] #save top 5 pathways
        tmp_df$pvalue = with(tmp_df, -log(pvalue,10)) #recalc pvalue
        tmp_df = tmp_df[order(tmp_df$ID),]
        names(tmp_df)[names(tmp_df) == "pvalue"] <- contrast_id #rename with contrast_id
        rownames(tmp_df) = NULL #remove rownames
        
        #save final df
        if(exists("final_pval")){
          final_pval = merge(final_pval,tmp_df,all=TRUE)
          final_pval[is.na(final_pval)] <- 0
        } else {
          final_pval=tmp_df
        }
        
        ## NES
        #pull top paths
        tmp_df = as.data.frame(tmp@result) #create df from results
        tmp_df = tmp_df[order(tmp_df$NES),] #order by pvalue
        tmp_df = tmp_df[1:5,c("ID","NES")] #save top 5 pathways
        tmp_df = tmp_df[order(tmp_df$ID),]
        names(tmp_df)[names(tmp_df) == "NES"] <- contrast_id #rename with contrast_id
        rownames(tmp_df) = NULL #remove rownames
        
        #save final df
        if(exists("final_nes")){
          final_nes = merge(final_nes,tmp_df,all=TRUE)
          final_nes[is.na(final_nes)] <- 0
        } else {
          final_nes=tmp_df
        }
      } else{
        print ("no significant genes were found, gsea not successful")
      }
      #increase deg counter
      counter_deg=counter_deg+1
    }

    #once all contrasts are complete, format db and save heatmaps
    if(count(complete.cases(final_pval),TRUE)<2){
      print("Heatmap cannot be generated with less than two sig pathways: pvalues")
    } else{
      print("--Saving heatplot for pvalues")
      
      #clean final df
      rownames(final_pval)=final_pval$ID #rownames as ID
      final_pval=final_pval[,2:ncol(final_pval)] #remove ID col
        
      #save fig
      p1 = pheatmap(final_pval,
                 color = colorRampPalette(c("white", "blue"))(256),
                 main = paste0("Heatplot of ", annotation_df[rownum,"file_name"], "\n pvalues"))
        ggsave(filename = paste0(out_dir, "pvalue/","heatplot.", annotation_df[rownum,"file_name"], ".png"),
               height= 8.90, width = 12.80, device = "png", plot = p1)
    }
    
    if(count(complete.cases(final_pval),TRUE)<2){
      print("Heatmap cannot be generated with less than two sig pathways: NES")
      } else{
        print("--Saving heatplot for NES")
        #clean final df
        rownames(final_nes)=final_nes$ID #rownames as ID
        final_nes=final_nes[,2:ncol(final_nes)] #remove ID col
        
        #save fig
        p1 = pheatmap(final_nes,
                 color = colorRampPalette(c("blue", "white"))(256),
                 main = paste0("Heatplot of ", annotation_df[rownum,"file_name"], "\n NES"))
        ggsave(filename = paste0(out_dir, "NES/","heatplot.", annotation_df[rownum,"file_name"], ".png"),
               height= 8.90, width = 12.80, device = "png", plot = p1)
      }
    
    #remove prev df
    remove(final_pval,final_nes)
  }
}

#set output
out_dir = paste0(analysis_dir,"heatplot/")

#set annotations                       
annotation_df = data.frame(c("C2","C2","C2","C2","C5","C5","C5"),
                           c("BIOCARTA","KEGG","REACTOME","WIKIPATHWAYS","GO:MF","GO:BP","GO:CC"),
                           c("c2b","c2k","c2r","c2w","c5mf","c5bp","c5cc"))
colnames(annotation_df) = c("ID","SUB","file_name")

#create heatplots with pvalues and NES values
gsea_plus_heatplot(contrast_list = compare_list,
                   deg_list = deg_list,
                   out_dir = out_dir,
                   annotation_df = annotation_df)
```

#####################################
# ORA on KMeans clusters
#####################################
#generate kmeans clusters at http://bioinformatics.sdstate.edu/idep/
#input /data/CCBR/projects/ccbr1109/analysis/deg_data/merged_deg_fc_DESeq2_20210811.txt
#output K means data buttton, saved to /data/CCBR/projects/ccbr1066/analysis/kmeans/Kmeans_bygroup.csv

Kmeans Functions
```{r}
kmeans_data_prep<-function(kmeans.path,cluster.group,df.in,version.id,out.dir){
  
  #read in kmeans data
  km_df = read.csv(kmeans.path)
  km_df = subset(km_df,Cluster==cluster.group)
  km_df = km_df[,c("Eid","Symbol",paste0(gsub("[.]","",compare_list),"fc"))]
  km_df[is.na(km_df)] <- ""
  
  #add variant info, pvalue to keg output
  for(rowid in rownames(km_df)){
    #pull ids from the merged df either the Ensembl with (complete) or without (partial) version number
    search_eid = subset(df.in,eid_complete==km_df[rowid,"Eid"])$gene[1]
    search_eid2 = subset(df.in,eid_partial==km_df[rowid,"Eid"])$gene[1]
    
    #set lookup column
    if(is.na(search_eid)){
      lookup_key = "eid_partial"
    } else{
      lookup_key = "eid_complete"
    }
    
    #add variant id 
    km_df[rowid,"Variant"] = subset(df.in,get(lookup_key)==km_df[rowid,"Eid"])$gene[1]
    
    for (contrast_id in compare_list){
      pvalue = min(subset(df.in, get(lookup_key)==km_df[rowid,"Eid"] & contrast==contrast_id)$pvalue)
      fdr = min(subset(df.in, get(lookup_key)==km_df[rowid,"Eid"] & contrast==contrast_id)$fdr)
  
      #if there is a pvalue, add it to the df
      if(length(pvalue)!=0){
        km_df[rowid,paste0(contrast_id,".p")] = pvalue
      }
      #if there is an fdr value, add it to the df
      if(length(fdr)!=0){
        km_df[rowid,paste0(contrast_id,".fdr")] = fdr
      }
    }
  }

  #cleanup symbol / variant
  for(rowid in rownames(km_df)){
    #if the symbol and variant are the same, remove variant
    if(km_df[rowid,"Symbol"]==km_df[rowid,"Variant"]){
      km_df[rowid,"Variant"] = ""
      
      #if there is no symbol, add variant to symbol
    } else if (km_df[rowid,"Symbol"]==""){
      km_df[rowid,"Symbol"] = km_df[rowid,"Variant"]
      km_df[rowid,"Variant"] = ""
    }
  }
  
  #fix pvalue na
  km_df[is.na(km_df)] <- "1"
  
  #rename headers
  for (contrast_id in compare_list){
    names(km_df)[names(km_df) == paste0(gsub("[.]","",contrast_id),"fc")] <- paste0(contrast_id,".fc")
  }
  
  #reorder columns
  km_df_out = km_df[,c(1, 2, length(compare_list)+3,
                3:(length(compare_list)+2),
                (length(compare_list)+4):ncol(km_df))]
  
  for (contrast_id in compare_list){
    km_df_sig = subset(km_df_out,get(paste0(contrast_id,".p"))<0.05)
  }
  
  #save files
  out_path = paste0(out.dir,"cluster_",cluster.group)
  print(paste0("Saving output to ",out_path))
  dir.create(file.path(out_path), showWarnings = FALSE)
  
  write.csv(km_df_out,paste0(out_path,"/kmeans_stats_all_",version.id,".csv"))
  write.csv(km_df_sig,paste0(out_path,"/kmeans_stats_sig_",version.id,".csv"))
  
  #return df
  return(km_df_out)
}

ora_function<-function(deg.file,out.dir,contrast.id, cluster.id){
  
  # read deg
  deg=deg.file
  
  #create output dir if needed
  dir.create(file.path(out.dir), showWarnings = FALSE)

  #subset significant genes for GSEA
  siggenes=deg[deg$fdr <= p_val & (deg$fc < (-1*fc) | deg$fc > fc),]
  sigGeneList=siggenes$gene
  
  #Run C2 plots
  c_list = c("BIOCARTA","KEGG","REACTOME","WIKIPATHWAYS")
  i=1
  print("C2 annotations")
  p1 = gsea_ora_plot(gl=sigGeneList,
                       t2g=annotate_genes("c2",c_list[i]),
                       ttl=paste0("ORA:C2:",c_list[i]," - ",contrast.id),
                        fn=paste0(out.dir,contrast.id,".c2.",c_list[i],".ora.results.txt",sep=""),
                       "ora")
  p2 = gsea_ora_plot(gl=sigGeneList,
                         t2g=annotate_genes("c2",c_list[i+1]),
                         ttl=paste0("ORA:C2:",c_list[i+1]," - ",contrast.id),
                          fn=paste0(out.dir,contrast.id,".c2.",c_list[i+1],".ora.results.txt",sep=""),
                         "ora")
  p3 = gsea_ora_plot(gl=sigGeneList,
                         t2g=annotate_genes("c2",c_list[i+2]),
                         ttl=paste0("ORA:C2:",c_list[i+2]," - ",contrast.id),
                          fn=paste0(out.dir,contrast.id,".c2.",c_list[i+2],".ora.results.txt",sep=""),
                         "ora")
  p4 = gsea_ora_plot(gl=sigGeneList,
                         t2g=annotate_genes("c2",c_list[i+3]),
                         ttl=paste0("ORA:C2:",c_list[i+3]," - ",contrast.id),
                          fn=paste0(out.dir,contrast.id,".c2.",c_list[i+3],".ora.results.txt",sep=""),
                         "ora")
  save_plots(p1,p2,paste0(out.dir,"c2.dotplot.png"),p3,p4)
  
  
  #Run C5 plots
  c_list = c("GO:MF","GO:BP","GO:CC")
  print("C5 annotations")
  p1 = gsea_ora_plot(gl=sigGeneList,
                       t2g=annotate_genes("c5",c_list[i]),
                       ttl=paste0("ORA:C5:",c_list[i]," - ",contrast.id),
                        fn=paste0(out.dir,contrast.id,".c5.",c_list[i],".ora.results.txt",sep=""),
                       "ora")
    p2 = gsea_ora_plot(gl=sigGeneList,
                         t2g=annotate_genes("c5",c_list[i+1]),
                         ttl=paste0("ORA:C5:",c_list[i+1]," - ",contrast.id),
                          fn=paste0(out.dir,contrast.id,".c5.",c_list[i+1],".ora.results.txt",sep=""),
                         "ora")
    p3 = gsea_ora_plot(gl=sigGeneList,
                         t2g=annotate_genes("c5",c_list[i+2]),
                         ttl=paste0("ORA:C5:",c_list[i+2]," - ",contrast.id),
                          fn=paste0(out.dir,contrast.id,".c5.",c_list[i+2],".ora.results.txt",sep=""),
                         "ora")
    save_plots(p1,p2,paste0(out.dir,"c5.dotplot.png"),p3)
}

kmeans_analysis<-function(){
  #created merged df of all degs
  counter=1
  for (deg in deg_list){
    tmp_df = read.csv(deg,sep="\t")[,c("ensid_gene","pvalue","fdr")]
    
    #separate eid and gene
    tmp_df = tmp_df %>%
      separate(ensid_gene, c("eid_complete", "gene"), "\\|")
  
    #separate eid and version
    tmp_df$eid_tmp = tmp_df$eid_complete
    tmp_df = tmp_df %>%
      separate(eid_tmp, c("eid_partial", "version"), "\\.")
  
    tmp_df$contrast = compare_list[counter]
    counter=counter+1
    
    #save final df
    if(exists("merged_df")){
        merged_df = merge(merged_df,tmp_df,all=TRUE)
      } else {
        merged_df=tmp_df
      }
  }

  #for each cluster (IE A,B,C from kmeans input)
  for (cluster_id in cluster_list){
    print(paste0("Starting with ",cluster_id))
    
    #merge kmeans with degs
    km_means_out = kmeans_data_prep(kmeans_input,cluster_id,merged_df,version_id,out_dir)
  
    #run ORA plots for each contrast
    for (contrast_id in compare_list){
      #subset df for specific contrast
      km_df = km_means_out[,c("Symbol",paste0(contrast_id,".fc"),paste0(contrast_id,".p"),paste0(contrast_id,".fdr"))]
      
      #rename subset to match expected input into ora
      km_df = km_df %>%
        rename(
          fc = paste0(contrast_id,".fc"),
          pvalue = paste0(contrast_id,".p"),
          fdr = paste0(contrast_id,".fdr"),
          gene= "Symbol"
        )
    
      print("**************")
      print(contrast_id)
      ora_function(deg.file = km_df, 
         out.dir = paste0(out_dir,"cluster_",cluster_id,"/",contrast_id,"_",version_id,"/"),
         contrast.id = contrast_id,
         cluster.id = cluster_id)
    }
  }
}
```

Kmeans Run
```{r}
#output dir
out_dir = "~/../../Volumes/ccbr1109/analysis/kmeans/"

kmeans_analysis()
```